### PR TITLE
KAFKA-15776: Support added to update remote.fetch.max.wait.ms dynamically

### PR DESCRIPTION
**Pull request by kamalcph**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
This patch also depends on #16199

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala [hunk 1]
```diff
@@ -792,6 +792,36 @@ class DynamicBrokerConfigTest {
...
+    val config = KafkaConfig(props)
+    val kafkaBroker = mock(classOf[KafkaBroker])
+    when(kafkaBroker.config).thenReturn(config)
+
+    val dynamicRemoteLogConfig = new DynamicRemoteLogConfig(kafkaBroker)
+    config.dynamicConfig.initialize(None, None)
+    config.dynamicConfig.addBrokerReconfigurable(dynamicRemoteLogConfig)
+
+    val newProps = new Properties()
+    newProps.put(RemoteLogManagerConfig.REMOTE_FETCH_MAX_WAIT_MS_PROP, "30000")
```
Comment (showuon) [inline core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:808]: We might need to assert the default value of `REMOTE_FETCH_MAX_WAIT_MS_PROP` before any change, to make sure the dynamically change takes effect.

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala [hunk 2]
```diff
@@ -792,6 +792,36 @@ class DynamicBrokerConfigTest {
...
+    // update default config
+    config.dynamicConfig.validate(newProps, perBrokerConfig = false)
```
Comment (showuon) [inline core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:810]: nit: assertDoesNotThrow

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala [hunk 3]
```diff
@@ -792,6 +792,36 @@ class DynamicBrokerConfigTest {
...
+    config.dynamicConfig.updateDefaultConfig(newProps)
+    assertEquals(30000, config.remoteFetchMaxWaitMs)
+
+    // update per broker config
+    newProps.put(RemoteLogManagerConfig.REMOTE_FETCH_MAX_WAIT_MS_PROP, "10000")
+    config.dynamicConfig.validate(newProps, perBrokerConfig = true)
```
Comment (showuon) [inline core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:816]: nit: assertDoesNotThrow

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala [hunk 4]
```diff
@@ -792,6 +792,36 @@ class DynamicBrokerConfigTest {
...
+    config.dynamicConfig.updateBrokerConfig(0, newProps)
+    assertEquals(10000, config.remoteFetchMaxWaitMs)
+
+    // invalid value
+    newProps.put(RemoteLogManagerConfig.REMOTE_FETCH_MAX_WAIT_MS_PROP, "-1")
```
Comment (showuon) [inline core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala]: Could we also test `0` for invalid case? Thanks.

**Review by showuon — COMMENTED** [review]
Thanks for the PR. Left some minor comments.

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala (hunk 2 above)
Comment (kamalcph) [reply core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:810]: the test will fail if the validation fails

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala (hunk 3 above)
Comment (kamalcph) [reply core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:816]: the test will fail if the validation fails

**Review by showuon — APPROVED** [review]
LGTM!

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala [hunk 5]
```diff
@@ -792,6 +792,40 @@ class DynamicBrokerConfigTest {
...
+    config.dynamicConfig.updateBrokerConfig(0, newProps)
+    assertEquals(10000, config.remoteFetchMaxWaitMs)
+
+    // invalid value "-1"
+    newProps.put(RemoteLogManagerConfig.REMOTE_FETCH_MAX_WAIT_MS_PROP, "-1")
+    assertThrows(classOf[ConfigException], () => config.dynamicConfig.validate(newProps, perBrokerConfig = true))
+    assertThrows(classOf[ConfigException], () => config.dynamicConfig.validate(newProps, perBrokerConfig = false))
+    // invalid value "0"
+    newProps.put(RemoteLogManagerConfig.REMOTE_FETCH_MAX_WAIT_MS_PROP, "0")
+    assertThrows(classOf[ConfigException], () => config.dynamicConfig.validate(newProps, perBrokerConfig = true))
```
Comment (satishd) [inline core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala]: Can you also check `perBrokerConfig = false` for value 0 like you did for negative value -1 in the earlier code?

**Review by satishd — COMMENTED** [review]
Thanks @kamalcph for the PR, overall LGTM. Left a minor comment in the test.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for addressing the review comments. LGTM.

**Comment by satishd** [comment]
A few test failures that are unrelated to this change, will merge it to trunk and 3.8 branches.

**Comment by satishd** [comment]
@kamalcph Can you raise a PR against 3.8 branch as I would like those changes to be reviewed in PR as it has conflicts with that branch?

**Comment by kamalcph** [comment]
> @kamalcph Can you raise a PR against 3.8 branch as I would like those changes to be reviewed in PR as it has conflicts with that branch?

Opened #16275 and #16276 to cherry-pick the KAFKA-15776 feature to the 3.8 branch. PTAL.

**Pull request merged**